### ISSUE TITLE
Block reward

### ISSUE THREAD

**Issue by cloudhead** [oscoin/oscoin#539]
To reward miners, we need to mint some tokens in every block and transfer them to a designated address, chosen by the miner. The way Ethereum does this, which is sub-optimal, is to include a "beneficiary" address in the block header, to which the block reward as well as tx fees are to be transfered. This is sub-optimal simply because the information is irrelevant to consensus, and thus light clients do not need it, but will have to download it anyway.

The way Bitcoin does it is better, but only really works with the UTXO model, where they have a transaction with no inputs as the first transaction of every block, and any "change" remaining in other transaction is considered a fee.

So how should we implement this? At the transaction level, it would complicate the semantics of transaction evaluation quite a bit, since the first transaction would provide "out of band" information to the following transactions, and validation would be more complicated too.

At the header level, we waste 32 bytes, and it's also not "correct".

A third option would be to have a kind of "intrinsics" section of the block that comes after the header and before the transactions, which would include data added by the miners, and could be later extended to do things other than block rewards.
```
Block { blockHeader :: Header, blockIntrinsics :: [Intrinsic], blockData :: [Tx] }
```
The only downside of this is that it complicates the block format.

Thoughts?

@kim @geigerzaehler @adinapoli-mndc

**Comment by adinapoli-mndc** [comment]
> At the transaction level, it would complicate the semantics of transaction evaluation quite a bit, since the first transaction would provide "out of band" information to the following transactions, and validation would be more complicated too.

I am not opposed to the "block intrinsics" proposal per se, and I do agree that storing this in the header is wasteful. However, let me play the devil's advocate on the statement above (it's early in the morning and I had no ☕️ yet, so.. 😁 ): how much would it complicate the evaluation semantic? Cannot a reward transaction be seen as a specific instantiation of a general `Tx`? In other words, intuitively I would have thought that rewards are (almost) indistinguishable from normal transactions: sure, they will receive special treatment as the _payer_ of the reward is _not_ another user/account (and that's the biggest question to answer, as we are not using an UTXO model), but for the rest this could be in principle a normal transaction that pays into a target `Account` no? (I know I am being a bit naive, but the important is to kickstart the conversation).

I think that if we make an effort in spelling out how hard would be to implement this as a normal(ish) transaction in the block body, I suspect this would be my preference, but if there is a strong technical reason not to go there, then I don't see any other way if not storing this into this new field, as doing that in the header is very (as you pointed out) wasteful.

**Comment by cloudhead** [comment]
I'll try to articulate better the complications of it being a transaction:
1. Since you also have to pay tx fees to this address, the "reward" tx would have to store the reward address somewhere for the subsequent txs to find.
2. Since the reward tx has to always and only be the first tx, tx application becomes context sensitive - if a "reward" tx is found later in the tx list, it needs to know that it isn't first in the list. Another way to do this would be to scan the list before evaluating the block, but that seems quite inefficient with thousands of txs.

**Comment by kim** [comment]
The "beneficiary address" approach costs 47 bytes per block iiuc, is that so bad?

**Comment by kim** [comment]
Also, iiuc, we'd be reordering the tx list anyways for parallelism, so a construction like `tx :| [tx]`, where the head is maybe a reward tx (and any subsequent ones are simply rejected) does not seem too bad either.

**Comment by cloudhead** [comment]
As it stands, the header will likely be ~148 bytes, so an increase of 47 bytes is quite a bit. This is only a problem for light clients, as 47 bytes is indeed nothing compared to the full block size including txs.

However, if we end up implementing something like FlyClient[0], then we are comparing on average log2(148) with log2(195), which is negligible. :thinking: 

[0]: https://eprint.iacr.org/2019/226

**Comment by cloudhead** [comment]
> Also, iiuc, we'd be reordering the tx list anyways for parallelism, so a construction like `tx :| [tx]`, where the head is maybe a reward tx (and any subsequent ones are simply rejected) does not seem too bad either.

Yeah, there might be a nice way to express this with a non-empty, and separate evaluation of the coinbase tx with the rest...

**Comment by cloudhead** [comment]
I'm going to give the NonEmpty list idea a try and let's see!

**Comment by cloudhead** [comment]
I started trying this out - it's not looking too bad so far, however there's one pretty annoying thing: there is no longer such thing as a "default genesis" block, because we need a coinbase transaction in there, and that requires a public key. So creating test data becomes more complicated. We can of course get around this by not using NonEmpty, and instead just checking this at runtime.

**Comment by adinapoli-mndc** [comment]
> there is no longer such thing as a "default genesis" block, because we need a coinbase transaction in there, and that requires a public key.

I mean, it's conceivable that in future we might want to have a different genesis block depending on the `network / environment` anyway, as opposed to a single static file inside `data/genesis.yaml`, no? In which case this is simply forcing us into design this properly, which might not necessarily be a bad thing.

**Comment by kim** [comment]
> one pretty annoying thing

Hm. We could donate the genesis coinbase :)

What I had in mind was not literally `NonEmpty`, but more like:

```haskell
data TxList tx = Maybe CoinbaseTx :| [tx]
```

Assuming that it is perfectly fine to not have a coinbase tx, in which case something needs to be burned I guess. Otherwise:

```haskell
data TxList tx = CoinbaseTx :| [tx]
```

Of course, we can also go crazy and say you cannot submit transactions including the coinbase, but you need to supply one for mining:

```haskell
data TxList f tx = f (CoinbaseTx) :| [tx]

type ClientSubmittedTxs tx = forall a. TxList (Const a) tx
type TxsForMining tx = TxList Identity tx
```

**Comment by cloudhead** [comment]
Ah I see, that is closer to what we came up with after discussing with @geigerzaehler - which is actually closer to the separate object solution:

```haskell
data Block c tx s = Block
    { blockHeader      :: BlockHeader c s
    , blockHash        :: BlockHash c
    , blockData        :: BlockData
    } deriving Generic

data BlockData c tx = BlockData
    { blockDataBeneficiary :: Maybe (PublicKey c)
    , blockDataTxs         :: [tx]
    }
```

One of the issues with having it as part of the tx list that came up is that it would have to be signed, have a nonce and fees etc. which obviously don't make sense if the miner is including it for themselves.

We can also quite easily make the beneficiary non-optional.
